### PR TITLE
Measurement Scaling

### PR DESCRIPTION
**Pull request by AarC10**
# Description
Add an option to scale measurements in case we want to decrease the size of our downlinks, while trying to maintain some precision

# How Has This Been Tested?
Ran FSW simulations and ran the backplane and receiver configurations, comparing the values seen in the new telemetry viewer application.

# Checklist:
- [ ] New functionality is documented in the necessary spots (i.e new functions documented in the header)
- [ ] Unit tests cover any new functionality or edge cases that the PR was meant to resolve (if applicable) 
- [ ] The CI checks are passing
- [ ] I reviewed my own code in the GitHub diff and am sure that each change is intentional

### CODE REVIEW

**Comment by cowsed** [comment]
Whats the one qodana warnign? it doesnt let me seewhat it is

File: lib/tlm/tlm_interpreter.go [hunk 1]
```diff
@@ -9,11 +9,12 @@ import (
...
-	Size       int    `yaml:"size"`                 // Size of the measurement in bytes
-	Type       string `yaml:"type,omitempty"`       // Type of the measurement (int, float)
-	Unsigned   bool   `yaml:"unsigned,omitempty"`   // Whether the measurement is unsigned
-	Endianness string `yaml:"endianness,omitempty"` // Endianness of the measurement (big, little)
+	Name       string  `yaml:"name"`                 // Name of the measurement
+	Size       int     `yaml:"size"`                 // Size of the measurement in bytes
+	Type       string  `yaml:"type,omitempty"`       // Type of the measurement (int, float)
+	Unsigned   bool    `yaml:"unsigned,omitempty"`   // Whether the measurement is unsigned
+	Endianness string  `yaml:"endianness,omitempty"` // Endianness of the measurement (big, little)
+	Scaling    float64 `yaml:"scaling,omitempty"`    // Scaling factor for the measurement (optional)
```
Comment (cowsed) [inline lib/tlm/tlm_interpreter.go:17]: Might want to add a comment (here and/or in FSW) that the scaling factor is multiplied here and divided by in FSW. I was trying to think of a more self explanatory name that implies divide on send, multiply on receive but i couldnt think of anything but scaling seems like it could go either way.

**Review by cowsed — COMMENTED** [review]
Yipee :partying_face:

File: lib/tlm/tlm_interpreter.go (hunk 1 above)
Comment (cowsed) [reply lib/tlm/tlm_interpreter.go:17]: Although, i suppose the scaling happens in code in FSW rather than in Config like it does here so maybe it doesnt matter that much. either way, it should probably be documented in a yaml comment or a go comment somewhere

**Comment by AarC10** [comment]
> Whats the one qodana warnign? it doesnt let me seewhat it is

image

Its all your fault man

**Comment by AarC10** [comment]
image

GitHub diff should show it as well

File: lib/tlm/tlm_interpreter.go [hunk 2]
```diff
@@ -97,17 +98,52 @@ func InterpretFloat(data []byte, endianness string) (interface{}, error) {
...
+		result, err = InterpretSignedInteger(data, measurement.Endianness)
 	case "float":
-		return InterpretFloat(data, measurement.Endianness)
+		result, err = InterpretFloat(data, measurement.Endianness)
 	default:
 		return nil, fmt.Errorf("unsupported type for measurement: %s", measurement.Type)
 	}
+
+	if measurement.Scaling != 1.0 {
+		switch v := result.(type) {
```
[the remark this answers is not in the document]
Comment (AarC10) [reply lib/tlm/tlm_interpreter.go:116]: Yes that is intentional. We don't know how big the actual number is and regardless, the scaling factor is a float too.

File: lib/tlm/tlm_interpreter.go (hunk 1 above)
Comment (AarC10) [reply lib/tlm/tlm_interpreter.go:17]: Im just gonna rename Scaling to say ScalingFactor. Factor should signify multiplication, not division

**Comment by cowsed** [comment]
ah darn it is my fault. when i tried to look it was complaining i dont have a cool jetbrains account

~~tbf you said i could ignore the CI stuff in my pr :stuck_out_tongue:~~

File: lib/tlm/tlm_interpreter.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (apenza) [reply lib/tlm/tlm_interpreter.go:116]: Please combine the cases

[the remark this answers is not in the document]
Comment (apenza) [reply lib/tlm/tlm_interpreter.go:116]: Wait, why can't we just always cast to a float64 and handle the error if one occurs?

**Review by apenza — CHANGES_REQUESTED** [review]

**Review by cowsed — APPROVED** [review]
looks good man
